### PR TITLE
API: Deal with params that are in deltas

### PR DESCRIPTION
**Pull request by jannotti**
For assets and apps, deals with holdings/locals changes being in deltas, but the params being in the db.

I originally opening the branch to make a `goal account applicationdetails` command to exercise paging, so that's in here too, since it allows nice e2e tests of the functionality that was broken.

The "proof" that this PR is needed is that removing the waiting rounds from `goal-account-asset.sh` didn't work properly.  Asset params where not being returned for assets that appeared in the deltas.  This PR removes them.

## Test Plan

Existing tests, including the removal of transactions that advanced the round in e2e subs tests. Those transactions prevented the exercise of delta/db merging corner cases showing up.

Also added an e2e sub app test in the style of the asset test.

### CODE REVIEW

File: ledger/acctupdates.go [hunk 1]
```diff
@@ -1452,9 +1345,28 @@ func (au *accountUpdates) lookupAssetResources(addr basics.Address, assetIDGT ba
...
+				if deltaParams, paramsInDelta := deltaParamsResults[assetID]; paramsInDelta {
+					if !deltaParams.Deleted {
+						arwi.Creator = deltaCreatorResults[assetID]
+						arwi.AssetParams = deltaParams.Params
+					}
+				} else {
+					// explicitly lookup the params in the db. Use
+					// currentDBRound to not bother examining the deltas
+					cid := basics.CreatableIndex(assetID)
+					creator, ok, err := au.getCreatorForRound(currentDBRound, cid, basics.AssetCreatable, false)
```
Comment (gmalouf) [inline ledger/acctupdates.go]: We need to hold accountsMu for this call and the one on 1364 I believe - can pass `synchronized=true` in for both functions or grab the lock outside and release after 1364.

Comment (jannotti) [reply ledger/acctupdates.go]: I was thinking we still had it locked here. But I suppose you're right since this is after we unlocked it.

Comment (jannotti) [reply ledger/acctupdates.go]: Fixed, by avoiding these calls entirely, and going straight to the DB. That also allows us to confirm the dbround has not advanced. https://github.com/algorand/go-algorand/pull/6588/changes/6cf79dba20149ab81f8943b8cc11e16fab404db4

File: cmd/goal/account.go [hunk 2]
```diff
@@ -791,27 +825,89 @@ func printAccountAssetsInformation(address string, response model.AccountAssetsI
 			if asset.AssetParams.Name != nil {
 				_, name = unicodePrintable(*asset.AssetParams.Name)
 			}
-			fmt.Printf("    Name: %s\n", name)
+			fmt.Printf("    Asset name: %s\n", name)
```
Comment (jannotti) [inline cmd/goal/account.go:834]: These ticky-tack naming changes are to have consistency with what's printed by the `asset info` command.

File: ledger/acctupdates.go [hunk 3]
```diff
@@ -1603,16 +1662,82 @@ func (au *accountUpdates) lookupApplicationResources(addr basics.Address, appIDG
 					arwi.AppLocalState = d.State.LocalState
 				}
 				if !d.Params.Deleted && d.Params.Params != nil {
-					arwi.Creator = addr
 					if includeParams {
+						arwi.Creator = addr
 						arwi.AppResource.AppParams = d.Params.Params
 					}
```
Comment (cce) [inline ledger/acctupdates.go]: This section is now unnecessary? It seems like it is already handled below in `if includeParams` starting at 1670.

Also the setting `arwi.Creator` and `arwi.AppResource.AppParams` here seem to conflict with the new deleted handling logic in 1674-1675

Comment (cce) [reply ledger/acctupdates.go]: also, this is not in the asset version

File: ledger/acctupdates.go [hunk 4]
```diff
@@ -1504,22 +1547,37 @@ func (au *accountUpdates) lookupApplicationResources(addr basics.Address, appIDG
...
+				if rec.Addr != addr {
 					continue
 				}
-				deltaResults[rec.Aidx] = rec
-				if rec.State.Deleted || rec.Params.Deleted {
+				// DB will return records that match for either params OR
+				// locals, so we may need an extra record if either is deleted
+				// in deltas.
+				if rec.Params.Deleted || rec.State.Deleted {
 					numDeltaDeleted++
```
Comment (cce) [inline ledger/acctupdates.go]: in the asset version, there is dedup for app asset ID
```
					if _, ok := deltaHoldingResults[rec.Aidx]; !ok {
						deltaHoldingResults[rec.Aidx] = rec
						if rec.Holding.Deleted {
							numDeltaDeleted++
						}
```

File: ledger/acctupdates.go [hunk 5]
```diff
@@ -1555,25 +1613,26 @@ func (au *accountUpdates) lookupApplicationResources(addr basics.Address, appIDG
...
 					als := pd.Data.GetAppLocalState()
 					arwi.AppLocalState = &als
 				}
 
-				if inDelta && d.Params.Deleted {
+				if paramsInDelta && deltaParams.Deleted {
 					// Delta deleted params — omit creator and params.
-				} else if inDelta && d.Params.Params != nil {
+				} else if paramsInDelta && deltaParams.Params != nil {
 					arwi.Creator = pd.Creator
```
Comment (cce) [inline ledger/acctupdates.go]: in the asset version, this is deltaCreatorResults[assetID] but I guess the same

File: ledger/acctupdates.go (hunk 3 above)
Comment (jannotti) [reply ledger/acctupdates.go]: I think I agree it is not needed, the later "patch in" seems like the more complete solution now.  I will remove and retest.

But I don't folllow what you're saying about "seem to conflict with the new deleted handling logic in 1674-1675".

Those are the lines that have the _same_ setting logic that we just said takes the place of this earlier code.

Comment (cce) [reply ledger/acctupdates.go]: I guess that's true I just noticed it was different if clauses that led you to the same lines so it was hard to understand if they would conflict or not

File: daemon/algod/api/client/restClient.go [hunk 6]
```diff
@@ -581,22 +587,32 @@ func (client RestClient) RawAccountApplicationInformation(accountAddress string,
...
 	return
 }
 
 // RawAccountAssetsInformation gets account information about a particular account's assets, subject to pagination.
-func (client RestClient) RawAccountAssetsInformation(accountAddress string, next *string, limit *uint64) (response []byte, err error) {
+func (client RestClient) RawAccountAssetsInformation(accountAddress string, next string, limit uint64) (response []byte, err error) {
 	var blob Blob
-	err = client.getRaw(&blob, fmt.Sprintf("/v2/accounts/%s/assets", accountAddress), pageParams{next, limit})
+	err = client.getRaw(&blob, fmt.Sprintf("/v2/accounts/%s/assets", accountAddress), accountAssetsParams{next, limit})
 	response = blob
```
[the remark this answers is not in the document]
Comment (jannotti) [reply daemon/algod/api/client/restClient.go:616]: This is our internal "client" there are no external users.

File: ledger/acctupdates.go (hunk 4 above)
Comment (jannotti) [reply ledger/acctupdates.go]: I unified the handling so that `numDeltaDeleted` is incremented for a params delete or a holding/local delete for both cases.  That can result in an overcount, but the overcount is safe. We just get more from the db.

File: ledger/acctupdates.go (hunk 5 above)
Comment (jannotti) [reply ledger/acctupdates.go]: I converted to `deltaCreatorResults[assetID]` for consistency, that's surely correct.  I can't come up with a way `pd.Creator` can be wrong/missing, but it's clear that if we see the params in the delta, we should get the creator from the same place.

File: ledger/acctupdates.go [hunk 7]
```diff
@@ -1452,9 +1465,38 @@ func (au *accountUpdates) lookupAssetResources(addr basics.Address, assetIDGT ba
...
+							return nil, 0, err
+						}
+						if persistedResource.Round != currentDBRound {
+							resourceDbRound = persistedResource.Round
+							goto retryWait
+						}
+						resource := persistedResource.AccountResource()
+						arwi.AssetParams = resource.AssetParams
+					}
 				}
```
[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/acctupdates.go:1526]: Ok, I've done this.

File: ledger/acctupdates.go [hunk 8]
```diff
@@ -1603,16 +1662,82 @@ func (au *accountUpdates) lookupApplicationResources(addr basics.Address, appIDG
...
+							if err != nil {
+								return nil, 0, err
+							}
+							if persistedResource.Round != currentDBRound {
+								resourceDbRound = persistedResource.Round
+								goto retryWait
+							}
+							resource := persistedResource.AccountResource()
+							arwi.AppParams = resource.AppParams
+						}
```
[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/acctupdates.go:1747]: Done

File: cmd/goal/account.go [hunk 9]
```diff
@@ -555,22 +564,59 @@ var assetDetailsCmd = &cobra.Command{
...
+				reportInfof("NextToken: %s", token)
+				break
+			}
 		}
+	},
+}
+var applicationDetailsCmd = &cobra.Command{
+	Use:   "applicationdetails",
+	Short: "Retrieve information about the applications created or opted-into by the specified account, possibly including application metadata",
+	Long:  `Retrieve information about the applications created or opted-into by the specified account, possibly including application metadata`,
```
Comment (cce) [inline cmd/goal/account.go:593]: So if you don't set `--include-params` (default false) for this command you won't see apps you created, despite the help text here. but if you do set `--include-params` you'll see AppParams not just for the apps you created but all the apps you've opted into also.. I guess it makes me wonder if the behavior for `--include-params false` should be to include AppParams for apps you created, as well as the local state for all the apps you opted into, or otherwise to update the help text to be clear that you won't see the apps you've created.

**Review by cce — DISMISSED** [review]

File: ledger/acctupdates.go [hunk 10]
```diff
@@ -1588,28 +1675,113 @@ func (au *accountUpdates) lookupApplicationResources(addr basics.Address, appIDG
...
```
[the remark this answers is not in the document]
Comment (cce) [reply ledger/acctupdates.go:1747]: here is an update to the unit tests to pin down Creator values in the assertions, it fails
https://github.com/algorand/go-algorand/commit/3824a44bd6cfd03c1aca6fd477538022f86b7575
https://github.com/jannotti/go-algorand/compare/goal-pagination...cce:go-algorand:goal-pagination-deltacreator

[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/acctupdates.go:1747]: Ok, I have fixes in my local work that I'll push soon that pass all of these tests, except one, which I believe is wrong.

Your tests retain a test that app 2001 is excluded (so it wants a returned list of len 5).  But my fixes cause this app to be returned - even though it's includeParams=false, my fixes populate the creator, and therefore return 2001. I think this is correct, because you still want to see the app you created, even if includeParams=false.

Agree?

File: cmd/goal/account.go (hunk 9 above)
Comment (jannotti) [reply cmd/goal/account.go:593]: I decided that a reasonable compromise is that if you don't use `--include-params` you _don't_ get params for anything, including your own created apps, _but_ the response still contains those apps, with the creator set to you.  So you still learn what "your" apps are, but you don't get the large params.

File: ledger/acctupdates.go [hunk 11]
```diff
@@ -1583,33 +1670,125 @@ func (au *accountUpdates) lookupApplicationResources(addr basics.Address, appIDG
 					}
 				}
 
-				if arwi.AppLocalState != nil || arwi.AppParams != nil {
+				// unlike assets, we might have !includeParams, which means we
+				// must also check for !Creator.IsZero to see if we should include.
+				if arwi.AppLocalState != nil || arwi.AppParams != nil || !arwi.Creator.IsZero() {
```
Comment (cce) [inline ledger/acctupdates.go]: If a non-creator closes out of an app in the deltas, then
- localsInDelta && deltaLocals.State.Deleted is true
- arwi.AppLocalState is nil
- paramsInDelta false unless the creator did something with params in the deltas
- pd.Creator will be set by DB query
- !arwi.Creator.IsZero will be true here

so an app that a non-creator closed out of would be in the API results?

Comment (cce) [reply ledger/acctupdates.go]: and the same for assets, the DB will fill pd.Creator and lead to a holder that closes out in deltas and then it gets assigned to arwi.AssetParams?

Comment (jannotti) [reply ledger/acctupdates.go]: I agree, and I will make the recommended changes.  I wish copilot noticed that when _I_ asked for a review.

Comment (jannotti) [reply ledger/acctupdates.go]: I think in assets, it can be fixed by going from:
```
if arwi.AssetHolding != nil || arwi.AssetParams != nil {
	result = append(result, arwi)
}
```
to
```
if arwi.AssetHolding != nil {
	result = append(result, arwi)
}
```
Unlike apps, we don't have to worry about a asset created by `addr` with no holding, that's impossible. So we can't just not include based on params alone.

File: ledger/acctupdates.go [hunk 12]
```diff
@@ -1409,24 +1426,25 @@ func (au *accountUpdates) lookupAssetResources(addr basics.Address, assetIDGT ba
...
 					// Delta deleted params — omit creator and params.
-				} else if inDelta && d.Params.Params != nil {
-					arwi.Creator = pd.Creator
-					arwi.AssetParams = d.Params.Params
+				} else if paramsInDelta && deltaParams.Params != nil {
+					arwi.Creator = deltaCreatorResults[assetID]
+					arwi.AssetParams = deltaParams.Params
 				} else if !pd.Creator.IsZero() {
 					arwi.Creator = pd.Creator
 					ap := pd.Data.GetAssetParams()
```
Comment (cce) [inline ledger/acctupdates.go:1450]: This sets arwi.Creator and arwi.AssetParams if an asset holder closes out in deltas
then later `if arwi.AssetParams != nil` is true

File: ledger/acctupdates.go (hunk 11 above)
Comment (cce) [reply ledger/acctupdates.go]: yes, this seems right

File: ledger/acctupdates.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (cce) [reply ledger/acctupdates.go:1747]: Yes that is right, it was pinning down the wrong behavior there

**Review by algorandskiy — DISMISSED** [review]

File: ledger/acctdeltas_test.go [hunk 13]
```diff
@@ -4175,23 +4181,26 @@ func TestLookupAssetResourcesWithDeltas(t *testing.T) {
...
 
 	// 1003: params updated in delta, holding preserved from DB
 	require.Equal(t, uint64(1003*100), assetMap[basics.AssetIndex(1003)].AssetHolding.Amount)
 	require.NotNil(t, assetMap[basics.AssetIndex(1003)].AssetParams)
 	require.Equal(t, uint64(7777), assetMap[basics.AssetIndex(1003)].AssetParams.Total)
 	require.Equal(t, "A1003new", assetMap[basics.AssetIndex(1003)].AssetParams.UnitName)
+	require.Equal(t, creatorAddr, assetMap[basics.AssetIndex(1003)].Creator)
 
 	// 1004: params deleted in delta, holding preserved from DB, no creator
 	require.Equal(t, uint64(1004*100), assetMap[basics.AssetIndex(1004)].AssetHolding.Amount)
```
Comment (cce) [inline ledger/acctdeltas_test.go]: this is a weird test case -- the creator deletes its params but keeps its holdings (which is impossible), and the lookup/API response has no creator and no params, since they were deleted but does show the holdings

File: ledger/acctupdates.go [hunk 14]
```diff
@@ -1583,33 +1674,125 @@ func (au *accountUpdates) lookupApplicationResources(addr basics.Address, appIDG
 					}
 				}
 
-				if arwi.AppLocalState != nil || arwi.AppParams != nil {
+				// unlike assets, we might have !includeParams, which means we
+				// must also check for addr == creator (if addr != creator, we don't care about it)
+				if arwi.AppLocalState != nil || arwi.AppParams != nil || arwi.Creator == addr {
```
Comment (cce) [inline ledger/acctupdates.go]:
```suggestion
				if arwi.AppLocalState != nil || arwi.Creator == addr {
```
I think this needs the same change as 97d6049eede2c6f53087e18441cda75c2de5b8fc

File: ledger/acctupdates.go [hunk 15]
```diff
@@ -1583,33 +1674,125 @@ func (au *accountUpdates) lookupApplicationResources(addr basics.Address, appIDG
...
+								goto retryWait
+							}
+							resource := persistedResource.AccountResource()
+							arwi.AppParams = resource.AppParams
+						}
 					}
 				}
-				if arwi.AppLocalState != nil || arwi.AppParams != nil {
+				// again, we must check for the creator, since that might be the only info here
+				if arwi.AppLocalState != nil || arwi.AppParams != nil || arwi.Creator == addr {
```
Comment (cce) [inline ledger/acctupdates.go]:
```suggestion
				if arwi.AppLocalState != nil || arwi.Creator == addr {
```
similar here, I think the same thing you did for assets and AssetParams needs to happen here

File: ledger/acctdeltas_test.go (hunk 13 above)
Comment (cce) [reply ledger/acctdeltas_test.go]: (didn't come from your PR but another example of existing tests asserting impossible stuff like where you fixed other tests)

File: ledger/acctupdates.go (hunk 14 above)
Comment (cce) [reply ledger/acctupdates.go]: For example: https://github.com/cce/go-algorand/commit/172964339500559fbe5a457d11ed0f6ceee4e3d1

File: ledger/acctdeltas_test.go (hunk 13 above)
Comment (jannotti) [reply ledger/acctdeltas_test.go]: I removed it in the least invasive way I could. Just did nothing to it, then expected it to be in the query.

File: ledger/acctupdates.go (hunk 14 above)
Comment (jannotti) [reply ledger/acctupdates.go]: agree

File: ledger/acctupdates.go (hunk 15 above)
Comment (jannotti) [reply ledger/acctupdates.go]: agree.

**Review by cce — APPROVED** [review]
LGTM will follow up with another PR to update tests maybe

**Review by algorandskiy — APPROVED** [review]
